tracer/src/Datadog.Trace/Debugger/Helpers/AsyncHelper.cs

```diff
@@ -0,0 +1,252 @@
+// <copyright file="AsyncHelper.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using Datadog.Trace.Debugger.Instrumentation;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.Debugger.Helpers
+{
+    internal static class AsyncHelper
+    {
+        internal static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(AsyncHelper));
+
+        private const string StateMachineNamePrefix = "<";
+        private const string StateMachineNameSuffix = ">";
+        private const string StateMachineFieldsNamePrefix = "<>";
+        private const string ThisName = "__this";
+        private const string StateMachineThisFieldName = StateMachineFieldsNamePrefix + "4" + ThisName;
+        // Should be the same as in debugger_tokens.h managed_profiler_debugger_is_first_entry_field_name
+        private const string StateMachineLiveDebuggerAddedField = "<>dd_liveDebugger_isReEntryToMoveNext";
+        private const BindingFlags AllFieldsBindingFlags = BindingFlags.Instance | BindingFlags.Static | BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.FlattenHierarchy;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static FieldInfo[] GetHoistedArgumentsFromStateMachine<TTarget>(TTarget instance, string[] originalMethodParametersName)
+        {
+            if (originalMethodParametersName is null || originalMethodParametersName.Length == 0)
+            {
+                return Array.Empty<FieldInfo>();
+            }
+
+            var foundFields = new FieldInfo[originalMethodParametersName.Length];
+            var allFields = instance.GetType().GetFields(AllFieldsBindingFlags);
+
+            int found = 0;
+            for (int i = 0; i < allFields.Length; i++)
+            {
+                // we have more fields than parameters, so we break once we found all parameters
+                if (found == foundFields.Length)
+                {
+                    break;
+                }
+
+                var field = allFields[i];
+                // early exit here based on generated name pattern, the assumption will be tested in exploration testing
```

Comment:
This is partly a TODO comment in disguise, so I'm suggesting a rephrase.
Also - can you verify this assumption now? 
```suggestion
                // If we can just guess the name based on the C# compiler generated name pattern, then we're done
```

Reply:
Fixed in 5a103fbe57cedbc9e13568d4372c82a921c46e26